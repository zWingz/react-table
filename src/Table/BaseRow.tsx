import * as React from 'react'
import { HTMLAttributes } from 'react'
import { PlainObject, ColumnProps } from './module'

type RowProp = Partial<HTMLAttributes<HTMLTableRowElement>>

export function getChainObject(obj: PlainObject, path: string) {
  const keys = path.split('.')
  if(keys.length === 1) {
    return obj[path]
  }
  const len = keys.length
  let tra = obj
  for (let i = 0; i < len; i++) {
    const key = keys[i]
    const tmp = tra[key]
    if (i === len - 1) {
      return tmp
    }
    if (tmp === undefined) {
      return undefined
    }
    tra = tmp
  }
}

interface TableRowProp<T> {
  record: T
  columns: ColumnProps<T>[]
  rowIndex: number
  onRow?: (record: T) => RowProp
}

// TODO: use React.memo instead of PureComponent when enzyme support

export default class TableRow<T extends PlainObject = any> extends React.PureComponent<TableRowProp<T>> {
  render() {
    const { record, onRow, columns, rowIndex } = this.props
      const trProp: RowProp = {}
      if (onRow) {
        const onRowProp = onRow(record) || {}
        Object.assign(trProp, onRowProp)
      }
      return (
        <tr {...trProp}>
          {columns.map((column, columnIdx) => {
            let value = null
            const { dataIndex, render, key } = column
            if (dataIndex) {
              if ((dataIndex as string).includes('.')) {
                value = getChainObject(record, dataIndex as string)
              } else {
                value = record[dataIndex as string]
              }
            }
            if (render) {
              value = render(value, record, rowIndex)
            }
            return (
              <td
                style={{ textAlign: column.align }}
                className={column.className}
                key={
                  key || (dataIndex as string) || columnIdx.toString()
                }
              >
                {value}
              </td>
            )
          })}
        </tr>
      )

  }
}

// export default React.memo(function TableRow<T extends PlainObject = any>(props: TableRowProp<T>) {
// }
// )
