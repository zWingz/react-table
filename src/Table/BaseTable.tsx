import React from 'react'
import classnames from 'classnames'
import PropTypes from 'prop-types'
import { ColumnProps } from './module'

interface BaseTableProp<T> {
  columns?: ColumnProps<T>[]
  dataSource?: T[]
  rowKey?: string
  top?: string | number,
  getRef?: (ref: any) => void,
  className?: string
  style?: React.CSSProperties
}

class BaseTable<T = any> extends React.PureComponent<BaseTableProp<T>> {
  static propTypes = {
    dataSource: PropTypes.array,
    columns: PropTypes.array,
    rowKey: PropTypes.string,
    top: PropTypes.number
  }
  static defaultProps = {
    getRef: () => {},
    style: {},
    dataSource: [],
    columns: []
  }
  get theadStyle(): React.CSSProperties {
    const { top } = this.props
    if(top) {
      return {
        transform: `translate3d(0px, ${top}px, 1px)`
      }
    }
    return {}
  }

  renderThead() {
    const { columns, top } = this.props
    return (
      <thead className={top ? 'fixed' : ''} style={this.theadStyle}>
        <tr>
          {columns.map((each, idx) => (
            <th key={idx}>{each.title}</th>
          ))}
        </tr>
      </thead>
    )
  }

  renderTbody() {
    const { columns, dataSource, rowKey } = this.props
    return (
      <tbody>
        {dataSource.map((record, idx) => {
          const keyValue = record[rowKey] + ''
          return (
            <tr key={keyValue}>
              {columns.map((column, columnIdx) => {
                let value = null
                const { dataIndex, render, key } = column
                if (dataIndex) {
                  value = record[dataIndex]
                }
                if (render) {
                  value = render(value, record, idx)
                }
                return (
                  <td
                    style={{textAlign: column.align}}
                    key={key || (dataIndex as string) || keyValue + columnIdx}
                  >
                    {value}
                  </td>
                )
              })}
            </tr>
          )
        })}
      </tbody>
    )
  }

  render() {
    return (
        <table ref={el => this.props.getRef(el)} className={classnames('fixed-table', this.props.className)} style={this.props.style}>
          {this.renderThead()}
          {this.renderTbody()}
        </table>
    )
  }
}

export default BaseTable
