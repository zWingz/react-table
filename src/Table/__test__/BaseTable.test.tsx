import * as React from 'react'
import { shallow, mount } from 'enzyme'
import BaseTable from '../BaseTable'
import BaseRow from '../BaseRow'
import { ColumnProps } from '../module'
import {default as dataSource} from './fixtures/DataSource'

type TestDataType = ColumnProps<typeof dataSource[0]>[]
const columns: TestDataType = [{
  title: 'id',
  dataIndex: 'id'
}, {
  title: 'createTime',
  dataIndex: 'createTime'
}, {
  title: <span>jsx title</span>,
  dataIndex: 'name'
}, {
  title: 'title',
  dataIndex: 'title',
  className: 'th-custom-class'
}]

describe('test BaseTable render', () => {
  const wrapper = shallow(<BaseTable columns={columns} dataSource={dataSource} rowKey='id'/>)
  it('test snapshot render', () => {
    // expect(wrapper.find(BaseRow)).toHaveLength(dataSource.length)
    expect(wrapper).toMatchSnapshot('baseTable snapshot')
  })
  it('test thead render', () => {
    const thead = wrapper.find('thead')
    expect(thead).toMatchSnapshot('thead snapshot')
    const th = thead.find('th')
    expect(th).toHaveLength(4)
    const ar = [0, 1, 3]
    ar.forEach((each) => {
      expect(th.at(each).text()).toEqual(columns[each].title)
    })
    expect(th.at(2).html()).toEqual('<th><span>jsx title</span></th>')
  })
  it('test tbody render', () => {
    expect(wrapper.find(BaseRow)).toHaveLength(4)
  })
})

describe('test BaseTable props', () => {
  let top = 0
  const className = 'custom-table', style = {
    marginTop: '10px',
    fontSize: '26px',
    padding: '10px 10px'
  }
  const wrapper = shallow(<BaseTable top={top} className={className} style={style} columns={columns} dataSource={dataSource} rowKey='id'/>)
  it('test classname', () => {
    expect(wrapper.hasClass('custom-table')).toBeTruthy()
    expect(wrapper.hasClass('fixed-table')).toBeTruthy()
  })
  it('test style', () => {
    expect(wrapper.find('table').props().style).toEqual(style)
  })
  it('test top', () => {
    let thead = wrapper.find('thead')
    expect(thead.hasClass('fixed')).toBeFalsy()
    top = 100
    wrapper.setProps({top: 100})
    thead = wrapper.find('thead')
    expect(thead.hasClass('fixed')).toBeTruthy()
    expect(thead.props().style.transform).toEqual(`translate3d(0px, ${top}px, 1px)`)
  })
})

// test later
// because enzyme not support React.memo

describe('test ref', () => {
  it('test ref', () => {
    const ref = jest.fn(val => val)
    const wrapper = mount(<BaseTable getRef={ref} columns={columns} dataSource={dataSource} rowKey='id'/>)
    expect(ref).toBeCalledTimes(1)
    const table = wrapper.getDOMNode()
    expect(ref.mock.calls[0][0]).toBe(table)
  })
})
