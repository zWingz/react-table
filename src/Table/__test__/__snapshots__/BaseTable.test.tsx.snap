// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test BaseTable render test snapshot render: baseTable snapshot 1`] = `
<table
  className="fixed-table"
  style={Object {}}
>
  <thead
    className=""
    style={Object {}}
  >
    <tr>
      <th
        key="0"
      >
        id
      </th>
      <th
        key="1"
      >
        createTime
      </th>
      <th
        key="2"
      >
        <span>
          jsx title
        </span>
      </th>
      <th
        className="th-custom-class"
        key="3"
      >
        title
      </th>
    </tr>
  </thead>
  <tbody>
    <TableRow
      columns={
        Array [
          Object {
            "dataIndex": "id",
            "title": "id",
          },
          Object {
            "dataIndex": "createTime",
            "title": "createTime",
          },
          Object {
            "dataIndex": "name",
            "title": <span>
              jsx title
            </span>,
          },
          Object {
            "className": "th-custom-class",
            "dataIndex": "title",
            "title": "title",
          },
        ]
      }
      key="440000200106140115"
      record={
        Object {
          "createTime": "1976-04-02",
          "id": "440000200106140115",
          "name": "尹娟",
          "title": "目派与响清",
        }
      }
      rowIndex={0}
    />
    <TableRow
      columns={
        Array [
          Object {
            "dataIndex": "id",
            "title": "id",
          },
          Object {
            "dataIndex": "createTime",
            "title": "createTime",
          },
          Object {
            "dataIndex": "name",
            "title": <span>
              jsx title
            </span>,
          },
          Object {
            "className": "th-custom-class",
            "dataIndex": "title",
            "title": "title",
          },
        ]
      }
      key="130000199202167395"
      record={
        Object {
          "createTime": "1978-12-17",
          "id": "130000199202167395",
          "name": "程丽",
          "title": "无北共外",
        }
      }
      rowIndex={1}
    />
    <TableRow
      columns={
        Array [
          Object {
            "dataIndex": "id",
            "title": "id",
          },
          Object {
            "dataIndex": "createTime",
            "title": "createTime",
          },
          Object {
            "dataIndex": "name",
            "title": <span>
              jsx title
            </span>,
          },
          Object {
            "className": "th-custom-class",
            "dataIndex": "title",
            "title": "title",
          },
        ]
      }
      key="640000197411167747"
      record={
        Object {
          "createTime": "2012-12-06",
          "id": "640000197411167747",
          "name": "何桂英",
          "title": "周切程",
        }
      }
      rowIndex={2}
    />
    <TableRow
      columns={
        Array [
          Object {
            "dataIndex": "id",
            "title": "id",
          },
          Object {
            "dataIndex": "createTime",
            "title": "createTime",
          },
          Object {
            "dataIndex": "name",
            "title": <span>
              jsx title
            </span>,
          },
          Object {
            "className": "th-custom-class",
            "dataIndex": "title",
            "title": "title",
          },
        ]
      }
      key="440000197608270476"
      record={
        Object {
          "createTime": "1990-07-18",
          "id": "440000197608270476",
          "name": "黄杰",
          "title": "一和下",
        }
      }
      rowIndex={3}
    />
  </tbody>
</table>
`;

exports[`test BaseTable render test thead render: thead snapshot 1`] = `
<thead
  className=""
  style={Object {}}
>
  <tr>
    <th
      key="0"
    >
      id
    </th>
    <th
      key="1"
    >
      createTime
    </th>
    <th
      key="2"
    >
      <span>
        jsx title
      </span>
    </th>
    <th
      className="th-custom-class"
      key="3"
    >
      title
    </th>
  </tr>
</thead>
`;
